### ISSUE TITLE
Unit Testing

### ISSUE THREAD

**Issue by dansiegel** [rotorgames/Rg.Plugins.Popup#256]
This is a wonderful library. But AFAIK, there is no way to currently do a Mock Initialization similar to how you can do a Mock Initialization of Xamarin Forms for Unit Testing.

It would be great if we could get a Mock Platform Initialization for Unit Testing.

**Comment by dansiegel** [comment]
@rotorgames I'm on the latest pre release. What I was thinking was if we had something like this added to PopupNavigation. It wouldn't show up in intellisense, but would make it easier to unit test with a mock instance. It would also make the plugin more extensible so that other platforms could be supported.

```cs
public static class PopupNavigation
{
    [EditorBrowsable(EditorBrowsableState.Never)]
    public static void SetInstance(IPopupNavigation instance) => 
        _popupNavigation = instance;
}
```

**Comment by rotorgames** [comment]
@dansiegel I don't think that `SetInstance` is needed. If you would like to use unit tests you can implement an instance of `IPopupNavigation` only for tests.

For example:

```csharp
// it is your view model which you would like to test
public class MyViewModel
{
  readonly IPopupNavigation _popupNavigation;

  public MyViewModel(IPopupNavigation popupNavigation)
  {
    _popupNavigation = popupNavigation;
  }
}
```

```csharp
// you can create the instance of vm for production
var vm = new MyViewModel(Rg.Plugins.Popup.Services.PopupNavigation.Instance);

//or you can create the instance of vm for unit tests
var vmForTests = new MyViewModel(new MyImplementationOfIPopupNavigation());
```

it is easy, we should not invoke `SetInstance` because it is not safe.

**Comment by dansiegel** [comment]
@rotorgames if that was the sort of testing I was doing then sure that would work. As you can see [here](https://github.com/dansiegel/Prism.Plugin.Popups/blob/master/src/Prism.Plugin.Popups/PopupRegistrationExtensions.cs#L8-L10) not having some way of initializing PopupNavigation is forcing me to create a hack substitution based on a build configuration. Either of the solutions I mentioned above would provide a much better way to properly initialize for testing.

**Comment by rotorgames** [comment]
@dansiegel I don't think that it is a problem. You can dot this:

```csharp
using System;
using System.Reflection;
using Prism.Behaviors;
using Prism.Ioc;
using Prism.Navigation;
using Rg.Plugins.Popup.Contracts;
using Rg.Plugins.Popup.Services;

namespace Prism.Plugin.Popups
{
    public static class PopupRegistrationExtensions
    {
        public static void RegisterPopupNavigationService<TService>(this IContainerRegistry containerRegistry)
            where TService : PopupPageNavigationService
        {
#if TEST
            containerRegistry.RegisterInstance<IPopupNavigation>(new Prism.Plugin.Popups.PluginNavigationMock());
#else
            containerRegistry.RegisterInstance<IPopupNavigation>(PopupNavigation.Instance);
#endif
            containerRegistry.RegisterSingleton<IPageBehaviorFactory, PopupPageBehaviorFactory>();
            if (IsDryIocContainer(containerRegistry))
            {
                containerRegistry.Register<INavigationService, TService>();
            }

            containerRegistry.Register<INavigationService, TService>(PrismApplicationBase.NavigationServiceName);
        }

        public static void RegisterPopupNavigationService(this IContainerRegistry containerRegistry) =>
            containerRegistry.RegisterPopupNavigationService<PopupPageNavigationService>();

        private static bool IsDryIocContainer(IContainerRegistry containerRegistry)
        {
            var regType = containerRegistry.GetType();
            var propInfo = regType.GetRuntimeProperty("Instance");
#if NETSTANDARD1_0
            return propInfo?.PropertyType.FullName.Equals("DryIoc.IContainer", StringComparison.OrdinalIgnoreCase) ?? false;
#else
            return propInfo?.PropertyType.FullName.Equals("DryIoc.IContainer", StringComparison.InvariantCultureIgnoreCase) ?? false;
#endif
        }
    }
}
```

I think it is better solution than:

```scharp
#if TEST
using PopupNavigation = Prism.Plugin.Popups.PluginNavigationMock;
#endif
```

Also you can see that there is not `SetInstance` in other plugins 

For example:
https://github.com/jamesmontemagno/PermissionsPlugin/blob/master/src/Plugin.Permissions/CrossPermissions.shared.cs

**Comment by dansiegel** [comment]
What I have there is a hack and actually the rename is cleaner than what you're showing by explicitly calling the mock... the mock doesn't belong in that code base at all as it should be inside of the unit test project. 

Again another workable solution here is that you make the IPopupPlatform public instead of internal. That would allow me to have a mock platform that could be exported and picked up by your plugin for testing. It also makes the plugin more extensible so if someone wanted to add Gtk, or WPF support they could without having to wait on a PR to be merged. 

Also if you look at the Xamarin Forms codebase you'll notice that they have a lot of methods that are hidden from intellisense that are available for you to use for unit testing.

**Comment by rotorgames** [comment]
@dansiegel Ok I think that make IPopupPlatform as a public is a good idea. I will think about it. But it can be a problem also. If a developer invokes `DependencyService.Get<IPopupPlatform>(DependencyFetchTarget.NewInstance)` then it can be a problem. I made IPopupPlatform as internal because of it.

**Comment by dansiegel** [comment]
In that scenario it shouldn't affect the IPopupNavigation instance as it would just give them a different instance of IPopupPlatform, and there isn't a reason for anyone to use an instance of IPopupPlatform. It's ultimately why I suggested the non-browsable approach because it would allow replacing the default instance with a mock for testing.

**Comment by rotorgames** [comment]
@dansiegel Ok I understand you but can you give me any libraries on C# which use `SetInstance` for example?

**Comment by dansiegel** [comment]
@rotorgames I can't think of any off the top of my head... but then again you have to consider:

1. Most people aren't actually running tests....
1. Taking an approach where you set the EditorBrowsableState to Never, also means it's only discoverable if you look at code or its documented... and AFAIK none of those methods are documented for Xamarin Forms even though they are all over the codebase.

**Comment by dansiegel** [comment]
@rotorgames here is an example of a current plugin that lets you set the current instance: https://github.com/smstuebe/xamarin-fingerprint#mocking-in-unit-tests

**Comment by rotorgames** [comment]
It will be available in v1.6.x
